File: android/app/src/main/kotlin/com/metalichesky/voicenote/util/recognize/RecognizeResult.kt
package com.metalichesky.voicenote.util.recognize

import com.google.gson.Gson
import com.google.gson.annotations.Expose
import java.lang.Exception

class RecognizeResult {
    companion object {
//        private val voskApiDelimiterRegex = Regex("(( )*:( )*)")
//        private val voskApiBracketsRegex = Regex("(^(( )*\")|(\"( )*)\$)")
        private val gson = Gson()

        fun fromVoskApiResult(voskApiResult: String): RecognizeResult {
            return try {
                gson.fromJson(voskApiResult, RecognizeResult::class.java)
            } catch (ex: Exception) {
                RecognizeResult()
            }
        }
    }

    enum class Type(val typeName: String) {
        PARTIAL("partial"),
        TEXT("text"),
        UNKNOWN("unknown")
    }

    var partial: String? = null
    var text: String? = null
    val type: Type
    get() {
        return when {
            text != null -> {
                Type.TEXT
            }
            partial != null -> {
                Type.PARTIAL
            }
            else -> {
                Type.UNKNOWN
            }
        }
    }

    val anyResult: String?
    get() {
        return text ?: partial
    }

    fun toVoskApiResult(): String {
        return "\"${type.typeName}\" : \"${anyResult}\""
    }

    override fun toString(): String {
        return gson.toJson(this);
    }

    fun isEmpty(): Boolean {
        return anyResult == null || anyResult?.isEmpty() == true;
    }

    override fun equals(other: Any?): Boolean {
        return other is RecognizeResult &&
                other.type == type &&
                other.text == text &&
                other.partial == partial
    }

    override fun hashCode(): Int {
        var result = partial?.hashCode() ?: 0
        result = 31 * result + (text?.hashCode() ?: 0)
        return result
    }
}

File: lib/domain/entity/recognize_result.dart


import 'package:json_annotation/json_annotation.dart';

part 'recognize_result.g.dart';

enum RecognizeResultType {
  PARTIAL,
  TEXT,
  UNKNOWN
}

extension RecognizeResultTypeExtension on RecognizeResultType {
  String get typeName {
    switch (this) {
      case RecognizeResultType.PARTIAL:
        return "partial";
      case RecognizeResultType.TEXT:
        return "text";
      case RecognizeResultType.UNKNOWN:
        return "unknown";
      default:
        return "unknown";
    }
  }
}

@JsonSerializable()
class RecognizeResult {
  String? text = null;
  String? partial = null;
  RecognizeResultType get type {
    if (partial != null) {
      return RecognizeResultType.PARTIAL;
    } else if (text != null) {
      return RecognizeResultType.TEXT;
    } else {
      return RecognizeResultType.UNKNOWN;
    }
  }

  RecognizeResult({required this.text, this.partial});

  String? get anyResult {
    if (partial != null) {
      return partial;
    } else {
      return text;
    }
  }

  factory RecognizeResult.fromJson(Map<String, dynamic> json) =>
      _$RecognizeResultFromJson(json);

  Map<String, dynamic> toJson() => _$RecognizeResultToJson(this);
}